src/Storages/MergeTree/calculateBlockNumbersForRestoring.cpp

```diff
@@ -0,0 +1,482 @@
+                const auto & right = mutations[right_index];
+                return left.number < right.number;
+            });
+
+        apply_permutation(mutations, mutation_names_in_backup, std::move(permutation));
+    }
+
+    struct PartMutations
+    {
+        /// Index of the first mutation this part is applied to or should be applied to.
+        ///
+        /// For example, let's consider a MergeTree with six parts "all_1_2_1_8", "all_3_3_0", "all_4_4_0_5", "all_6_6_0_8", "all_7_7_0", "all_9_9_0"
+        /// and two mutations "mutation_5.txt", "mutation_8.txt". For those parts `first_mutation` and `num_applied_mutations` will be calculated as:
+        /// | part_name   | first_mutation | num_applied_mutations |
+        /// | all_1_2_1_8 | 0 (mutation_5) |                     2 |
+        /// | all_3_3_0   | 0              |                     0 |
+        /// | all_4_4_0_5 | 0              |                     1 |
+        /// | all_6_6_0_8 | 1 (mutation_8) |                     1 |
+        /// | all_7_7_0   | 1              |                     0 |
+        /// | all_9_9_0   | 2 (none)       |                     0 |
+        ///
+        size_t first_mutation = 0;
+
+        /// Number of mutations this part is already applied to.
+        size_t num_applied_mutations = 0;
+    };
+
+    /// Find corresponding mutations for each part.
+    std::vector<PartMutations>
+    findMutationsForEachPart(const std::vector<MergeTreePartInfo> & parts, const std::vector<MutationInfoFromBackup> & mutations)
+    {
+        std::vector<PartMutations> res;
+        res.resize(parts.size());
+
+        if (mutations.empty())
+        {
+            /// No mutations.
+        }
+        else if (mutations.front().block_number)
+        {
+            /// Mutations are specified with a single `block_number` per a mutation.
+
+            /// Make a map of mutations where mutations are ordered by their block numbers.
+            /// This map needs to be ordered because we use upper_bound() below.
+            std::map<Int64, size_t> mutations_by_block_number;
+
+            for (size_t i = 0; i != mutations.size(); ++i)
+            {
+                const auto & mutation = mutations[i];
+
+                /// MergeTree and ReplicatedMergeTree mutations can't be mixed.
+                if (!mutation.block_number || mutation.block_numbers)
+                    throw Exception(ErrorCodes::CANNOT_RESTORE_TABLE, "Mutations specify incorrect block numbers");
+                Int64 block_number = *mutation.block_number;
+
+                /// The block number specified in a mutation must increase with following mutations.
+                if (i > 0)
+                {
+                    Int64 previous_block_number = *mutations[i - 1].block_number;
+                    if (block_number <= previous_block_number)
+                        throw Exception(ErrorCodes::CANNOT_RESTORE_TABLE, "Mutations specify incorrect block numbers");
```

Comment:
It's better to use it because it makes the algorithm simpler - see the usage of `upper_bound()`, and also because it makes the algorithm more consistent with the algorithm for ReplicatedMergeTree.